cwms-data-api/src/main/java/cwms/cda/data/dao/JooqDao.java

```diff
@@ -422,16 +426,80 @@ private static InvalidItemException buildInvalidItem(RuntimeException input) {
             cause = dae.getCause();
         }
 
-        InvalidItemException exception = new InvalidItemException(cause);
+        String message = "Invalid Item.";
 
         String localizedMessage = cause.getLocalizedMessage();
         if (localizedMessage != null) {
             String[] parts = localizedMessage.split("\n");
             if (parts.length > 1) {
-                exception = new InvalidItemException(parts[0], cause);
+                message = parts[0];
             }
         }
-        return exception;
+        return new InvalidItemException(message, cause);
+    }
+
+    public static boolean isInvalidUnits(RuntimeException input) {
+        boolean retVal = false;
+
+        Optional<SQLException> optional = getSqlException(input);
+        if (optional.isPresent()) {
+            SQLException sqlException = optional.get();
+            String message = sqlException.getLocalizedMessage();
+            int errorCode = sqlException.getErrorCode();
+
+            retVal = errorCode == 20998
+                    && message.contains("ORA-20102: The unit")
+                    && message.contains("is not a recognized CWMS Database unit for the")
+                ;
+        }
+        return retVal;
+    }
+
+    private static InvalidItemException buildInvalidUnits(RuntimeException input) {
+
+        Throwable cause = input;
+        if (input instanceof DataAccessException) {
+            DataAccessException dae = (DataAccessException) input;
+            cause = dae.getCause();
+        }
+
+        String localizedMessage = cause.getLocalizedMessage();
+        if (localizedMessage != null) {
+            // skip ahead in localizedMessage to "ORA-20102:"
+            String searchFor = "ORA-20102:";
+            int start = localizedMessage.indexOf(searchFor);
+            if (start >= 0) {
+                localizedMessage = localizedMessage.substring(start + searchFor.length());
+                String[] parts = localizedMessage.split("\n");
+                if (parts.length >= 1) {
+                    localizedMessage = parts[0];
+                }
+            }
+        }
+
+        localizedMessage = sanitizeOrNull(localizedMessage);
+
+        if (localizedMessage == null || localizedMessage.isEmpty()) {
+            localizedMessage = "Invalid Units.";
+        }
+
+        return new InvalidItemException(localizedMessage, cause);
+    }
+
+    private static @Nullable String sanitizeOrNull(@Nullable String localizedMessage) {
```

Comment:
Couple of things. There is a sanitizer defined in ApiServlet, we should pass that around as it will then contain the project global settings.

Two, if sanitized, you can just pass the data back, `::sanitize` makes the data safe to return to the user. So setting null isn't required.

Reply:
Also to be fair if it's just a pain to get that sanitizer instance here, this is fine. But the sanitized data can be passed back to the user.

Reply:
It may be safe to return but it doesn't seem helpful.    
"The unit: <script>alert()</script> is not a recognized CWMS Database unit for the Temp Parameter"   becomes "The unit: "
Which the user would end up seeing as:
{
"message": "Bad Request.",
"incidentIdentifier": "-12341234123451234521345",
"details": {
"message": "The unit: "
}
}

I guess that is better than nothing and its only nefarious users that would see that message but it sort of looks like CDA is in the middle of a stroke.  Seemed better to just return something we control than to let them figure out where in the input our sanitizer started complaining.

Reply:
passing the ctx in isn't convenient b/c the exception get wrapped from jooq listener thing.  Not saying its impossible. I did see the sanitizer in ApiServlet and couldn't think of a clean way to get it.

Reply:
That's fair. Had to ask. Put a note in about that though so we all remain aware.

Reply:
> It may be safe to return but it doesn't seem helpful. "The unit: <script>alert()</script> is not a recognized CWMS Database unit for the Temp Parameter" becomes "The unit: " Which the user would end up seeing as: { "message": "Bad Request.", "incidentIdentifier": "-12341234123451234521345", "details": { "message": "The unit: " } }
> 
> I guess that is better than nothing and its only nefarious users that would see that message but it sort of looks like CDA is in the middle of a stroke. Seemed better to just return something we control than to let them figure out where in the input our sanitizer started complaining.

Hmm, yeah, and it's only checking for `<script>` which is always invalid anyways.